### ISSUE TITLE
File Recovery

### ISSUE THREAD

**Issue by fxha** [marktext/marktext#732]
### Describe your feature request

Implement a document recovery feature which can:

1. guard documents during saving against IO errors, application crashes and power loss ("bad" write: `open file handle (file is truncated to 0 bytes) --> crash --> new content would be written`)
2. create regular backups from changed files to prevent lost during power loss or shutdown

... controlled via settings.

### What problem does this feature solve? [optional]

Lost of documents and changes when Mark Text is terminated. This issue should be implemented together with a session to save the latest workspace (tabs, view state, etc).

**Comment by Jocs** [comment]
I understand that you need a historical version of the file. We regularly help the user keep a historical version of the edited text, for recovery after file loss and rollback to the historical version?

**Comment by Jocs** [comment]
I have also considered this feature, including some of its ancillary features, we can choose the historical version when rolling back, including the historical version created time, preview, and the diff contrast with the latest version (if there is the latest version), in the storage history In the version, we also need to delete the old historical version regularly to save memory. For example, we may save it in 3 minutes in the recent editing. The old history may be one version per day or  per month.

**Comment by fxha** [comment]
I thought about a simple session and recovery feature such as in vscode or Firefox. More specifically that we save the session and files (tabs) regularly and make Mark Text single instance. Then we can run a thread (or process via `electron-remote` because node doesn't support threads) to regularly (user-defined - lets say 60s) write changed tabs to the backup location and overwrite the old backup. Session changes should be written into a state, which is saved before the application is closing and regularly via the "recovery" thread. After a forced restart we can restore all tabs or notify the user about the problem. A forced restart can be detected for example in case a `recovery` file exists in the backup location. While the application is closing we remove this file and delete old backups in the background.

> The state of various window, tab and user-data will be saved, and reloaded upon application start. The feature will be passive, reacting only to forced-restarts such as extension installation and crashes. [...]
> Features
> 
>  - Save data for each window and tab (P1)
>  - Automatically restore the session data upon restarting from a forced closure (P1)
> -  If post-crash, allow the user to choose whether to restore (P1)
>  - Allow the user to "undo" closing a tab or window (P2)
>
> https://wiki.mozilla.org/Session_Restore

**Comment by lrq3000** [comment]
Please implement this feature, it's very critical. It's not a feature most people see at first, but when they need it, this is what will make them either migrate to another app (whether it provides more security or not), or stick and love this app.

I can't help in programming, but I can suggest to take inspiration from Notepad++, which has the most reliable session and autosave feature I have ever seen, it's highly robust, I have never lost any file with it. In fact, I everyday just close Notepad++ with it being full of unsaved documents/tabs, and I reopen the next day without any loss, and I do this for weeks/months. A nice feature is that it's possible to specify where the unsaved documents are stored, which I have pointed to a folder that I automatically back up on the cloud with SpiderOak. So even if Notepad++ fails (which has never happened yet), it's improbable I'll lose any document. That's the kind of flexibility their implementation allows, and it simply relies on autosaving regularly unsaved documents wherever the user points to.

**Comment by Jocs** [comment]
@Irq3000 Very good advice

**Comment by Jerakin** [comment]
I just lost a huge document and my hope brought me here only to be crushed by the [Open]. Needless to say I would love if this was implemented, and I use Notepad++ the same way that Irq3000 does so something similar would be optimal. But I think that if the app is terminated suddenly then it could try to save a "recovery" file at least and prompt the user on startup if they want to restore it.

**Comment by moshpirit** [comment]
I'm new to this project, so I don't really know how MT works but maybe the file could be auto-saved in a temporal directory (/tmp/...) every X seconds and if it fails, it can use this backed up file as a recovery file.

**Comment by lrq3000** [comment]
Yes that's exactly what Notepad++ does, to save every X seconds a backup of all opened files/tabs, and that's likely the safest way to provide a file recovery feature as otherwise any mechanism that only activates when the app crashes is: 1) more complex to implement as it requires a different process to run in parallel to the main app and detect when the main app crashed, 2) is prone to also crash or be unable to recover the files in case the crash is due to external circumstances (ie, power outage, OS crash, etc), in which cases only periodic saving can recover the files.

If this gets implemented, I would strongly recommend to allow the user to define what directory to save backups to, this allows to use cloud backup services to add some redundancy or even version history.

**Comment by YFlag** [comment]
Ahhhhhhhhh I just created lots of unsaved untilted tabs, with lots of draft text written and lots of images captured, due to convenience (Cuz there is often the case that fast information recording is needed, such as meeting, remote communication..); Afterwards I just covered my thinkpad and heartbreakingly, it fell into sleep and cannot be awakened any more!!! I have to force to restart it.... And Now the unsaved tabs are allllll lost !!!! 555555555..... I cannot live!!!! So saaaad o(╥﹏╥)o Ahhhhh 

So please, I beg you, please offer some functions, such as caching these unsaved documents, or recovering, like that... It's just for the peace of the world!!!!!! for the saving the lives of people like me...

**Comment by vimkim** [comment]
just happened... I lost so many things...

**Comment by Aliwasan** [comment]
I have just lost my file content and I don't understand why because I saved regularly (event hough MarkText is doing it automatically). It is a lot of work and I really need to know if I can recover it please help !

**Comment by vimkim** [comment]
I lost everything and it makes me use vscode again. Please implement this feature.

**Comment by Ar-Kareem** [comment]
This would be an amazing feature to have (and a must for me personally to switch to marktext).

I generally have 5-10 documents open at a time filled with content in addition to unnamed new documents not saved to disk yet. Whenever windows decides to be windows and auto-restart the PC for an update, I need to be able to recover my content or I can't justify using the software.

Notepad++ does this very reliably and it would be amazing to also have that reliability here.

**Comment by Youngshuang** [comment]
Does Marktext have logs like Typora? If I have not turned on autosave and Unexpected shutdown the computer without manually saving, how can I recover the content????????????????

**Comment by JFWooten4** [comment]
Just happened here for a few hrs of work

**Comment by DTBtrigger** [comment]
No!!!  I lost all my content I edited all day. 
Please tell me there is a way to recover them. T~T

**Comment by moshpirit** [comment]
I don't think this project is supported anymore, maybe we should start looking for alternatives.

**Comment by JFWooten4** [comment]
re @moshpirit, do you know what lead to the downtread?

**Comment by lrq3000** [comment]
There is Typora, which appeared around the same time as MarkText but has been continually maintained and updated to this day, but it's a paid app. It has auto save and recovesy by the way, as well at autocopy of internet images on copy-pasting, and rendering of diagrams such as mermaid. 

Unfortunately it takes a lot of time to work on a note taking editor, so I am not surprised MarkText got dropped at some point.

Furthermore IIRC it is based on electron, which it a dead technology nowadays, so even if someone was willing to continue the project, it would require significant rewriting that probably would make the endeavor meaningless as it would take less time to make from scratch a superior wysiwyg markdown editor using newer technos, especially now that we have AI assisted coding.

**Comment by memewizards** [comment]
why hasn't this been implemented yet? I just lost a week of notes.

**Comment by Neclyon** [comment]
MAMA I JUST LOST A HUNDRED-WORD BLOG
